File: edu/api/info.py
#  Nikulin Vasily © 2021
from flask_login import current_user

from data import db_session
from data.homeworks import Subject
from data.users import User
from edu.api import socket, api
from tools.tools import send_response


@socket.on('getSubjects')
@api.route('/api/subjects', methods=['GET'])
def getSubjects():
    event_name = 'getSubjects'

    db_sess = db_session.create_session()

    subjects = list(db_sess.query(Subject.title).all())

    return send_response(
        event_name,
        {
            'message': 'Success',
            'subjects': sorted(list(map(lambda x: x[0], subjects)))
        }
    )


@socket.on('getTeachersList')
@api.route('/api/teachers-info', methods=['GET'])
def getTeachersList():
    event_name = 'getTeachersList'

    db_sess = db_session.create_session()

    school_id = current_user.school_id
    teachers = list(filter(lambda user: user.has_role('teacher'),
                           db_sess.query(User).filter(
                               User.school_id == school_id
                           ).all()))

    teachers_data = [f'{t.surname} {t.name} {t.patronymic}' for t in teachers]

    return send_response(
        event_name,
        {
            'message': 'Success',
            'teachers': sorted(teachers_data)
        }
    )

File: edu/static/js/api/info.js
/*
 * Nikulin Vasily © 2021
 */

info = Object()
info.subjects = function (fn = null) {
    if (fn)
        info.subjectsFn = fn
    else
        info.subjectsFn = null
    socket.emit('getSubjects')
}

socket.on('getSubjects', function (data) {
    info.subjectsJson = data
    if (info.subjectsFn)
        info.subjectsFn()
})

info.teachersList = function (fn = null) {
    if (fn)
        info.teachersListFn = fn
    else
        info.teachersListFn = null
    socket.emit('getTeachersList')
}

socket.on('getTeachersList', function (data) {
    info.teachersListJson = data
    if (info.teachersListFn)
        info.teachersListFn()
})